File: julia/testLocalPhs.jl

using LinearAlgebra
using HaltonSequences

include("LocalPhs.jl")

################################################################################

# A script for testing the speed and accuracy of local polyharmonic spline (PHS)
# interpolation of scattered data in high dimensions (up to 7).

# Greg Barnett
# September 2023

################################################################################

dims = 5                                      # dimensions (1, 2, 3, 4, 5, 6, 7)
rbfExponent = 3                  # odd number exponent to use in the phs rbf (3)
polyDegree = 1                      # maximum polynomial degree in the basis (1)
a = -1;  b = 1                  # bounds on the domain (used for all dimensions)
nNodes = 100000                                                # number of nodes
nEvalPts = 100                                     # number of evaluation points
stencilRadius = 6/16        # how far away to look for neighbors of a given node
useHalton = false        # boolean to decide if you want to use halton or random

################################################################################

base1 = [2 5 11 17 23 31 41]            # base of primes for making halton nodes
base2 = [3 7 13 19 29 37 43]         # base of primes for making halton eval pts

# Get the nodes and evaluate the test function there.
nodes = rand(nNodes, dims)
if useHalton
    h = HaltonPoint(base1[1:dims], length=nNodes)
    for i = 1 : size(nodes, 1)
        nodes[i,:] = h[i]
    end
end
nodes = nodes * (b - a) .+ a
zzNodes = Phs_testFunc(nodes)

# Get the evaluation points and evaluate the test function there.
evalPts = rand(nEvalPts, dims)
if useHalton
    h = HaltonPoint(base2[1:dims], length=nEvalPts)
    for i = 1 : size(evalPts, 1)
        evalPts[i,:] = h[i]
    end
end
evalPts = evalPts * (0.80 * (b - a)) .+ (a + 0.10 * (b - a))
# evalPts = evalPts * (b - a) .+ a
zzEvalPts = Phs_testFunc(evalPts)
# println("minEvalPt = ", minimum(evalPts))
# println("maxEvalPt = ", maximum(evalPts))

# Use local polyharmonic spline interpolation to estimate values at evalPts.
phs = LocalPhs_new(rbfExponent, polyDegree, nodes, zzNodes, stencilRadius)
estimate = LocalPhs_evaluate(phs, evalPts)

# Check the error against the true underlying function at the evalPts.
diff = abs.(estimate[:] - zzEvalPts[:])
avgErr = norm(diff, 1) / length(diff)
println("average error = ", avgErr)
println("maximum error = ", norm(diff, Inf))
println("error std dev = ", norm(diff .- avgErr, 2) / (length(diff) - 1))

File: perl/testLocalPhs.pl
#!/usr/bin/perl

use strict;
use warnings;

use lib ".";
use LocalPhs;

my $rbfExponent = 3;
my $polyDegree = 1;

# Define spatial domain [a,b] x [c,d], with m rows and n columns.
my $a = -1;  my $b = 1;  my $n = 6;
my $c = -1;  my $d = 1;  my $m = 6;

# Define number of rows (M) and columns (N) for evaluation points.
my $M = 10;
my $N = 10;

# Define the size of the stencil.
my $stencilRadius = 1;

################################################################################

my ($x, $y, $xx, $yy, $nodes, $zz);
eval {
    $x = Matrix::linspace($a, $b, $n);
    $y = Matrix::linspace($c, $d, $m);
    ($xx, $yy) = Matrix::meshgrid($x, $y);
    $xx = $xx->flatten;  $yy = $yy->flatten;
    $nodes = $xx->vstack($yy)->transpose;
    $zz = Phs::testFunc2d($nodes);
};
die if $@;

my ($X, $Y, $XX, $YY, $NODES, $ZZ);
eval {
    $X = Matrix::linspace($a, $b, $N);
    $Y = Matrix::linspace($c, $d, $M);
    ($XX, $YY) = Matrix::meshgrid($X, $Y);
    $XX = $XX->flatten;  $YY = $YY->flatten;
    $NODES = $XX->vstack($YY)->transpose;
    $ZZ = Phs::testFunc2d($NODES);
};
die if $@;

my ($phs, $estimate);
eval {
    $phs = LocalPhs::new($rbfExponent, $polyDegree, $nodes, $zz, $stencilRadius);
    $estimate = $phs->evaluate($NODES);
};
die if $@;

my $diff;
eval {
    print "\$estimate = \n";
    $estimate->transpose->disp;
    print "\n";
    print "\$ZZ = \n";
    $ZZ->transpose->disp;
    print "\n";
    $diff = $estimate->plus($ZZ->dot(-1));
    print "\$diff = \n";
    $diff->transpose->disp;
};
die if $@;
